File: app/src/main/java/me/ykrank/s1next/view/adapter/delegate/ForumAdapterDelegate.java
package me.ykrank.s1next.view.adapter.delegate;

import android.content.Context;
import android.databinding.DataBindingUtil;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.ViewGroup;

import javax.inject.Inject;

import me.ykrank.s1next.App;
import me.ykrank.s1next.R;
import me.ykrank.s1next.data.api.model.Forum;
import me.ykrank.s1next.data.pref.ThemeManager;
import me.ykrank.s1next.databinding.ItemForumBinding;
import me.ykrank.s1next.viewmodel.ForumViewModel;

public final class ForumAdapterDelegate extends BaseAdapterDelegate<Forum, ForumAdapterDelegate.BindingViewHolder> {
    @Inject
    ThemeManager themeManager;

    public ForumAdapterDelegate(Context context) {
        super(context);
        App.getAppComponent().inject(this);
    }

    @NonNull
    @Override
    protected Class<Forum> getTClass() {
        return Forum.class;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent) {
        ItemForumBinding binding = DataBindingUtil.inflate(mLayoutInflater,
                R.layout.item_forum, parent, false);
        binding.setGentleAccentColor(themeManager.getGentleAccentColor());
        binding.setForumViewModel(new ForumViewModel());

        return new BindingViewHolder(binding);
    }

    @Override
    public void onBindViewHolderData(Forum forum, int position, @NonNull BindingViewHolder holder) {
        ItemForumBinding binding = holder.itemForumBinding;
        binding.getForumViewModel().forum.set(forum);
        binding.executePendingBindings();
    }

    static final class BindingViewHolder extends RecyclerView.ViewHolder {

        private final ItemForumBinding itemForumBinding;

        public BindingViewHolder(ItemForumBinding itemForumBinding) {
            super(itemForumBinding.getRoot());

            this.itemForumBinding = itemForumBinding;
        }
    }
}

File: app/src/main/java/me/ykrank/s1next/view/adapter/delegate/ForumAdapterDelegate.kt
package me.ykrank.s1next.view.adapter.delegate

import android.content.Context
import android.databinding.DataBindingUtil
import android.support.v7.widget.RecyclerView
import android.view.ViewGroup
import com.github.ykrank.androidtools.ui.adapter.simple.SimpleRecycleViewHolder
import me.ykrank.s1next.App
import me.ykrank.s1next.R
import me.ykrank.s1next.data.api.model.Forum
import me.ykrank.s1next.data.pref.ThemeManager
import me.ykrank.s1next.databinding.ItemForumBinding
import me.ykrank.s1next.viewmodel.ForumViewModel
import javax.inject.Inject

class ForumAdapterDelegate(context: Context) : BaseAdapterDelegate<Forum, SimpleRecycleViewHolder<ItemForumBinding>>(context, Forum::class.java) {
    @Inject
    internal lateinit var themeManager: ThemeManager

    init {
        App.appComponent.inject(this)
    }

    public override fun onCreateViewHolder(parent: ViewGroup): RecyclerView.ViewHolder {
        val binding = DataBindingUtil.inflate<ItemForumBinding>(mLayoutInflater,
                R.layout.item_forum, parent, false)
        binding.gentleAccentColor = themeManager.gentleAccentColor
        binding.forumViewModel = ForumViewModel()

        return SimpleRecycleViewHolder<ItemForumBinding>(binding)
    }

    override fun onBindViewHolderData(t: Forum, position: Int, holder: SimpleRecycleViewHolder<ItemForumBinding>, payloads: List<Any>) {
        val binding = holder.binding
        binding.forumViewModel?.forum?.set(t)
    }

}
